File: 3507-find-the-count-of-numbers-which-are-not-special/3507-find-the-count-of-numbers-which-are-not-special.go
func nonSpecialCount(l int, r int) int {
    lim := int(math.Sqrt(float64(r)))
	isPrime := make([]bool, lim+1)
	for i := 2; i <= lim; i++ {
		isPrime[i] = true
	}
	isPrime[0] = false
	isPrime[1] = false

	for i := 2; i*i <= lim; i++ {
		if isPrime[i] {
			for j := i * i; j <= lim; j += i {
				isPrime[j] = false
			}
		}
	}

	cnt := 0
	for i := 2; i <= lim; i++ {
		if isPrime[i] {
			square := i * i
			if square >= l && square <= r {
				cnt++
			}
		}
	}

	totalCount := r - l + 1
	return totalCount - cnt
}

File: 3507-find-the-count-of-numbers-which-are-not-special/3507-find-the-count-of-numbers-which-are-not-special.java
class Solution {
    public int nonSpecialCount(int l, int r) {
        int lim = (int) Math.sqrt(r);
        boolean isPrime[] = new boolean[lim+1];
        Arrays.fill(isPrime , true);
        isPrime[0] = isPrime[1] = false; 

        for(int i = 2; i*i <= lim ; i++){
            if(isPrime[i]){
                 for (int j = i * i; j <= lim; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        int cnt = 0 ;
        for(int i = 2; i<= lim ;i++){
             if (isPrime[i]) {
                int square = i * i;
                if (square >= l && square <= r) {
                   cnt++;
                }
        }
        }
         int totalCount = r - l + 1;
        return totalCount - cnt;
    }
}